core/engine.go

```diff
@@ -307,6 +318,8 @@ func (e *Engine) processThresholds(abort func()) {
 	}
 
 	if abortOnFail && abort != nil {
+		//TODO: When sending this status we get a 422 Unprocessable Entity
+		e.setRunStatus(lib.RunStatusAbortedThreshold)
```

Comment:
I believe there is an issue here. 

When this case is reached, it will fire off a call to`SetRunStatus` on the `Collector` which sends to a buffered channel size 1 (in the `cloud` implementation). Problem is that after this, there will be another call to `e.setRunStatus` [here](https://github.com/loadimpact/k6/pull/606/files#diff-cfd4da29308a83fc01fc28fae9be4bbaR224)  before anybody gets to read from the mentioned channel and the main goroutine will deadlock.

Reply:
Hi, @antekresicthanks and @na-- 

That line is not supposed to be there, I forgot it. There is no need to set that. This method is supposed to be called only once per run.

So, I would suggest the simplest approach, a single mutex protecting from a potential data-race, since this is being introduced in order to make `Insights` a bit more user-friendly.

The whole canceling stuff seems to be overcomplicated, but I guess part of that feeling is just the lack of our own knowledge of the code base and on how canceling on Go it's like. If you put some more time reading the code and trying to refactor, you will see that it will be hard to make that great.

Reply:
Hmm after giving it a bit of thought, it seems to me that either:
- we don't need a mutex or any synchronization at all, if a data race is impossible
- a data race is possible, but in that case just a simple mutex won't be enough, since we need to not only ensure data synchronization, but also the order of operations (setting the status should happen before reading it)

Reply:
Sure, if you're sure there are no data races, I haven't investigated the whole workflow thoroughly, so I'm not sure. 

If you want to be safe, since this will be somewhat difficult to test with the cloud collector, also implement `SetRunStatus()` in the `Dummy` collector to save the passed status as a property and read/log it after `<-ctx.Done()` in `Run()`. This way the tests would probably panic with `-race` if there's a race condition there.